hw/ip/otbn/dv/rig/README.md

```diff
@@ -0,0 +1,106 @@
+# OTBN Random Instruction Generator
+
+This directory contains a random instruction generator for OTBN called
+`otbn-rig`. This is intended to be run in multiple phases. If you just
+want to generate some random binaries, it might be easier to call the
+wrapper at `../uvm/gen-binaries.py`.
+
+At the moment, there are two sub-commands: `gen` and `asm`. In future,
+we might add more (to do things like test case mutation or shrinking).
+
+## The gen command
+
+The `gen` command is in charge of actually generating a random
+program. This program is written to stdout in an JSON format. This
+format may change over time, but it is understood by the `asm`
+command.
+
+Example usage:
+```
+hw/ip/otbn/dv/rig/otbn-rig gen --seed 123 --size 1000 >foo.json
+```
+
+To control random generation, there is a `--seed` parameter. The
+output should be stable for a fixed seed. If not specified, the seed
+is zero.
+
+To support testing with the `START_ADDR` register, which allows the
+host processor to control OTBN's reset address, there is a
+`--start-addr` parameter. By default this is zero.
+
+### The size parameter
+
+The `--size` parameter is used to control how big the program grows. A
+snippet with a single unconditionally executed instruction has a size
+of one. A snippet containing a sequence of `N` instructions has a size
+of `N`. For a more interesting example, consider an if/else branch
+that would look something like this in C:
+
+```C
+    if (A) {
+        B;
+    } else {
+        C;
+    }
+```
+
+If calculating `A` takes `a` cycles and `B` and `C` have a size of `b`
+and `c` respectively, then the entire snippet has size `a + max(b,
+c)`. The general idea is that the size taken by a snippet is an upper
+bound on the number of instructions that it causes the processor to
+execute. This might not be a strict upper bound: for example, it won't
+be strict if `a != b` in the if/else snippet above.
+
+### Snippets
+
+The random program is built from blocks called "snippets". The
+simplest possible snippet is a sequence of straight-line instructions.
+There are also more complicated snippets like conditional branches and
+loops. The JSON output format represents this tree of snippets.
+
+### Initialised data
+
+The generated program is designed never to trigger architecturally
+unspecified behaviour. It might trigger errors, but its execution will
+never depend on things like the initial contents of the register file
+or uninitialised memories.
+
+If there is no initialised data in the image, this means any load
```

Comment:
```suggestion
If there would be no initialised data in the image, this means any load
```

Or swap the paragraphs 2 and 3 to have the implemented functionality first, and the motivation after that.

Reply:
Yeah, I rewrote this paragraph three times trying to come up with a nice way to say things. I've just pushed up a new version that gives up and doesn't provide as much detail. Much easier to read, and anyone who wants to know more can read the comments in the code!